Bug 1435511 - Keep the WatchedRepo menu open unless one of the links is clicked

KWierso opened the pull request:
No matter where you clicked on the page, the menu was always being toggled shut.

I couldn't get stopPropagation() or preventDefault() to stop that from happening, so this PR just forces the toggle attributes to stay in their "show" mode unless one of the links in the menu is clicked. Happy to do this some other, less hacky way, if you're aware of one.

Local testing seems to still let the menu get closed if I click outside of the menu or click the toggle button itself.

camd commented:
@KWierso : I'm wondering if you'd be up for experimenting with ``reactstrap`` ``dropdowns`` : https://reactstrap.github.io/components/dropdowns/
I think if you set the item as a ``header`` then it won't close when you click on it.  At least the example on that page works that way.  It may slightly change the styling, but perhaps that's ok.

KWierso commented:
Here's what the old repo navbar looked like:
![screenshot_2018-10-05 1 mozilla-inbound bug 1178783 add crashtest for now-fixed flexbox hang bug no re](https://user-images.githubusercontent.com/172215/46566007-2ab21e80-c8cc-11e8-9553-80471b79931c.png)


And here's what the new one looks like:
![screenshot_2018-10-05 1 mozilla-inbound bug 1178783 add crashtest for now-fixed flexbox hang bug no re 1](https://user-images.githubusercontent.com/172215/46566006-2ab21e80-c8cc-11e8-8622-75c2d9d50488.png)
(Note that the "closed" background color applies to the buttons to the right of the closed repo's name)


The old dropdown:
![screenshot_2018-10-05 1 mozilla-inbound bug 1178783 add crashtest for now-fixed flexbox hang bug no re 2](https://user-images.githubusercontent.com/172215/46566005-2ab21e80-c8cc-11e8-9c13-5aa3c878c708.png)

And the new one:
![screenshot_2018-10-05 1 mozilla-inbound bug 1178783 add crashtest for now-fixed flexbox hang bug no re 3](https://user-images.githubusercontent.com/172215/46566004-2a198800-c8cc-11e8-9046-1cf543216d43.png)

I think I need to play with the classNames a bit more, when you hover over the dropdown items, the background of the item gets a doubled up 'hover' highlight:
![screenshot_2018-10-05 1 mozilla-inbound bug 1178783 add crashtest for now-fixed flexbox hang bug no re 4](https://user-images.githubusercontent.com/172215/46566070-fbe87800-c8cc-11e8-9977-7f4a321201f6.png)

But that's what I got before the weekend hit.

KWierso commented:
@edmorley @camd Any preference for how this should work?
If you click the repo name for the active repo, it just toggles the dropdown open/closed. If you do that but have filters applied, it briefly opens, then wipes out the filters. If you click for a repo other than the active one, it briefly opens the menu before navigating to the other repo.

I'd guess we want to not let the repo name toggle the dropdown.

KWierso commented:
Okay, I think this should be reviewable and landable.

Here's the navbar:
![screenshot_2018-10-08 31 mozilla-esr52](https://user-images.githubusercontent.com/172215/46639095-f3847d00-cb18-11e8-800d-2538d4f58702.png)

Here's the active repo's dropdown:
![screenshot_2018-10-08 31 mozilla-esr52 1](https://user-images.githubusercontent.com/172215/46639094-f2ebe680-cb18-11e8-8066-947daf1e98b6.png)

Here's an inactive repo that's open:
![screenshot_2018-10-08 31 mozilla-esr52 2](https://user-images.githubusercontent.com/172215/46639093-f2ebe680-cb18-11e8-9f93-e5db5d0194ae.png)

And an inactive repo for something not in treestatus:
![screenshot_2018-10-08 31 mozilla-esr52 3](https://user-images.githubusercontent.com/172215/46639092-f2ebe680-cb18-11e8-88ff-4620c64df1e2.png)

The doubled up hover effect is fixed, and clicking the repo name no longer briefly toggles the dropdown open before the page reloads.

camd commented on ui/job-view/headerbars/WatchedRepo.jsx:
```diff
@@ -119,65 +120,65 @@ export default class WatchedRepo extends React.Component {
           {status === 'unsupported' && <React.Fragment>
-            <li className="watched-repo-dropdown-item">
+            <DropdownItem header>
               <span>{watchedRepo} is not listed on <a
```
We don't need this surrounding ``<span>`` anymore.  Not sure if we ever did, tbh.  :)

camd commented on ui/job-view/headerbars/WatchedRepo.jsx:
```diff
@@ -119,65 +120,65 @@ export default class WatchedRepo extends React.Component {
+            rel="noopener noreferrer"
+            href={`https://mozilla-releng.net/treestatus/show/${treeStatusName}`}
+          >
+            <span>Tree Status</span>
```
Don't need this ``<span>`` element

camd commented on ui/job-view/headerbars/WatchedRepo.jsx:
```diff
@@ -119,65 +120,65 @@ export default class WatchedRepo extends React.Component {
+            rel="noopener noreferrer"
+            href={repo.pushLogUrl}
+          >
+            <span>Pushlog</span>
```
don't need this ``<span>``

camd commented on ui/job-view/headerbars/WatchedRepo.jsx:
```diff
@@ -119,65 +120,65 @@ export default class WatchedRepo extends React.Component {
-        ><span className="fa fa-times" /></button>}
-
-        <ul className="dropdown-menu" role="menu">
+        <DropdownToggle
```
This could just be a single line.

camd commented on ui/css/treeherder-navbar.css:
```diff
@@ -474,3 +486,7 @@ fieldset[disabled] .btn-view-nav-closed.active {
     background-color: #d3d3d34d;
 }
+
+.dropdown-item > span {
```
This doesn't seem to make any difference.  I don't think we need this CSS styling, do we?

KWierso replied:
This was to force the Treestatus and Pushlog dropdown items to stay black despite having visited the page. It doesn't apply now that I'm removing the superfluous <span>s elsewhere in this review. I can change the rule to `a.dropdown-item` to get the same effect.

KWierso replied:
superfluous SPANs*. Silly github parsing...

camd replied:
OK, cool.  sounds good.  :)

camd commented on ui/css/treeherder-navbar.css:
```diff
@@ -140,9 +140,21 @@ secondary-nav-bar {
+  max-width: 250px;
+}
+
+.watched-repo-menu > .dropdown-header {
```
Can you give me some context on why these CSS items are needed?  With a long message, it really DOES seem like we need what you have here for ``.watched-repo-menu > .dropdown-header``.  But the others don't really seem necessary.  But I might be missing their reasoning.  :)  Do you think we can do without them?

KWierso replied:
I can drop the extra word-wrap without a problem. The min/max-widths are to try to standardize the size of the dropdowns. Otherwise, repos with MotDs or closure reasons have menus that are much wider than repos without those messages. 

Happy to drop it all if that's fine with you.

camd replied:
If it ends up needing it, that's totally fine.  :)  I've just recently become aware of how bloated our ``.CSS`` is.  @sarah-clements had brought it up to me that we can get some good mileage out of https://getbootstrap.com/docs/4.1/utilities/sizing/ and the other utilities in there.  If you wouldn't mind, would you see if there's anything in there that can be used instead?  If not, no worries, do what you need to.  :)

sarah-clements commented:
@KWierso Does this pr still need to be open?

KWierso commented:
I don't see myself finishing this anytime soon. Feel free to close it out if you want.
